### PR TITLE
bpf_metadata: Use original source address even if the destination is external

### PR DESCRIPTION
**Pull request by jrajahalme**
Use the original source address if permitted even if the destination is external to the cluster.

### CODE REVIEW

File: cilium/bpf_metadata.cc [hunk 1]
```diff
@@ -384,12 +384,9 @@ bool Config::getMetadata(Network::ConnectionSocket& socket) {
     // This means that a local host IP is used if no IP is configured to be used instead of it
     // ('ip' above is null).
     src_address = nullptr;
-  } else if (!(use_original_source_address_ &&
-               !(destination_identity & Cilium::ID::LocalIdentityFlag) &&
-               destination_identity != Cilium::ID::WORLD && !npmap_->exists(other_ip))) {
-    // Otherwise only use the original source address if permitted, destination identity is not a
-    // locally allocated identity, is not classified as WORLD, and the destination is not in the
-    // same node.
+  } else if (!use_original_source_address_ || npmap_->exists(other_ip))) {
```
Comment (mhofstetter) [inline cilium/bpf_metadata.cc:387]: `||` -> `&&` ?

File: cilium/bpf_metadata.cc [hunk 2]
```diff
@@ -384,12 +384,9 @@ bool Config::getMetadata(Network::ConnectionSocket& socket) {
...
+    // Otherwise only use the original source address if permitted and the destination is not
+    //in the same node.
```
Comment (mhofstetter) [inline cilium/bpf_metadata.cc:389]: nit: `//in` -> `// in` (add space after slash)

File: cilium/bpf_metadata.cc (hunk 1 above)
Comment (jrajahalme) [reply cilium/bpf_metadata.cc:387]: I inverted the logic via de Morgan's laws to simplify the logic (outer `not` is applied to each element inside the brackets and each `and`/`or` is flipped to `or`/`and`).

Comment (mhofstetter) [reply cilium/bpf_metadata.cc:387]: ah right, i overlooked that it's not only the `use_original_source_address` that was inverted. I kind of gave the reason for this simplification just by asking and being able to read it correctly :smile:

File: cilium/bpf_metadata.cc (hunk 2 above)
Comment (jrajahalme) [reply cilium/bpf_metadata.cc]: fixed

File: cilium/bpf_metadata.cc [hunk 3]
```diff
@@ -384,12 +384,9 @@ bool Config::getMetadata(Network::ConnectionSocket& socket) {
...
     // ('ip' above is null).
     src_address = nullptr;
-  } else if (!(use_original_source_address_ &&
-               !(destination_identity & Cilium::ID::LocalIdentityFlag) &&
-               destination_identity != Cilium::ID::WORLD && !npmap_->exists(other_ip))) {
-    // Otherwise only use the original source address if permitted, destination identity is not a
-    // locally allocated identity, is not classified as WORLD, and the destination is not in the
-    // same node.
+  } else if (!use_original_source_address_ || npmap_->exists(other_ip)) {
+    // Otherwise only use the original source address if permitted and the destination is not
```
Comment (mhofstetter) [inline cilium/bpf_metadata.cc:388]: nit: the comment reads a little bit strange. `Otherwise` would more relate to the `else` case (instead of `else if`).

I would rephrase this to something like this (to help documenting the actual condition)

> Don't use the original source address if not configured or the destination is on the same node.

**Comment by jschwinger233** [comment]
Hi @jrajahalme , may I confirm as of this PR, proxy uses original src IP even for go-to-world traffic, like pod to 1.1.1.1 ?
